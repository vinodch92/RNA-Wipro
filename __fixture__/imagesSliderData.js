export const data = [
  {
    format: 'jpeg',
    width: 5616,
    height: 3744,
    filename: '0.jpeg',
    id: 0,
    author: 'Alejandro Escamilla',
    author_url: 'https://unsplash.com/photos/yC-Yzbqy7PY',
    post_url: 'https://unsplash.com/photos/yC-Yzbqy7PY',
  },
  {
    format: 'jpeg',
    width: 5616,
    height: 3744,
    filename: '1.jpeg',
    id: 1,
    author: 'Alejandro Escamilla',
    author_url: 'https://unsplash.com/photos/LNRyGwIJr5c',
    post_url: 'https://unsplash.com/photos/LNRyGwIJr5c',
  },
  {
    format: 'jpeg',
    width: 2500,
    height: 1667,
    filename: '10.jpeg',
    id: 10,
    author: 'Paul Jarvis',
    author_url: 'https://unsplash.com/photos/6J--NXulQCs',
    post_url: 'https://unsplash.com/photos/6J--NXulQCs',
  },
  {
    format: 'jpeg',
    width: 2500,
    height: 1656,
    filename: '100.jpeg',
    id: 100,
    author: 'Tina Rataj',
    author_url: 'https://unsplash.com/photos/pwaaqfoMibI',
    post_url: 'https://unsplash.com/photos/pwaaqfoMibI',
  },
];
