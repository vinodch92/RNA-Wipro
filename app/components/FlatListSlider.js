import React, { useCallback, useEffect, useRef } from 'react';
import { StyleSheet, FlatList, Dimensions, LayoutAnimation } from 'react-native';

/**
 * This functional component is FlatListSlider and useful to show images with author name on sliding.
 * @param {*} params : params consists of contentContainerStyle object for customizing style of flatlist
 * component, dataLength and ...props are all props of flatlist.
 */
const FlatListSlider = ({ contentContainerStyle, dataLength, ...props }) => {
  /* useRef for creating a reference of scrollview */
  let flatListRef = useRef(null);

  /* useRef for creating a reference of sliderTimer */
  let sliderTimer = useRef(null);

  /* useEffect hook to start the auto play of slides */
  useEffect(() => {
    startAutoPlay();
  }, [startAutoPlay]);

  /**
   * This function is responsible to change the current index of slider array in to local state
   */
  const changeSliderListIndex = useCallback(() => {
    LayoutAnimation.configureNext(LayoutAnimation.Presets.easeIn);

    let slideIndex = 1;
    if (flatListRef?.current?.index && flatListRef.current.index < dataLength - 1) {
      slideIndex = flatListRef.current.index + 1;
    }
    flatListRef.current.index = slideIndex;

    flatListRef.current.scrollToIndex({
      index: slideIndex,
      animated: true,
    });
  }, [dataLength]);

  /**
   * This function is responsible to set the timeInterval for starting the autoplay of slider
   */
  const startAutoPlay = useCallback(() => {
    if (dataLength > 0) {
      sliderTimer.current = setInterval(changeSliderListIndex, 3000);
    }
  }, [changeSliderListIndex, dataLength]);

  /* keyExtractor for flatlist */
  const keyExtractor = (item, index) => `${item.id}_${index}`;

  return (
    <FlatList
      ref={flatListRef}
      pagingEnabled={true}
      snapToInterval={Math.round(Dimensions.get('screen').width)}
      decelerationRate="fast"
      bounces={false}
      horizontal
      getItemLayout={(data, index) => ({
        length: Math.round(Dimensions.get('screen').width),
        offset: Math.round(Dimensions.get('screen').width) * index,
        index,
      })}
      windowSize={1}
      initialNumToRender={1}
      maxToRenderPerBatch={1}
      removeClippedSubviews={true}
      showsHorizontalScrollIndicator={false}
      {...props}
      contentContainerStyle={[styles.contentContainer, contentContainerStyle]}
      keyExtractor={keyExtractor}
    />
  );
};

const styles = StyleSheet.create({
  contentContainer: { alignItems: 'center' },
});

export default FlatListSlider;
