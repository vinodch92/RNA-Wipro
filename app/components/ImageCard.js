import React from 'react';
import { View, Text, Image, StyleSheet, Platform } from 'react-native';
import config from '../config';

/**
 * This functional component is ImageCard and useful to show images with author name for image slider.
 * @param {*} params : params consists of id of image, authorName, containerStyle object for customizing
 * main container, textStyle object for customizing authorName text style, imageWrapperStyle object for
 * customizing style of wrapper of image, textNumLines to specify the number of lines for Text element.
 */
const ImageCard = ({
  id = 0,
  authorName,
  containerStyle,
  textStyle,
  textNumLines,
  imageWrapperStyle,
}) => {
  const imageSource = `${config.imagesDataBaseUrl}/${config.imageWidth}/${config.imageHeight}?image=${id}`;

  return (
    <View key={id} style={[styles.container, containerStyle]}>
      {authorName ? (
        <Text
          style={[styles.text, textStyle]}
          numberOfLines={textNumLines}
          testID={config.accessibilityStrings.AUTHOR_NAME_TEXT}>
          {authorName}
        </Text>
      ) : null}
      <View style={[styles.imageWrapper, imageWrapperStyle]}>
        <Image style={styles.image} source={{ uri: imageSource }} />
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: { alignItems: 'center' },
  text: {
    fontSize: 16,
    fontWeight: config.fontWeights.FONT_WEIGHT_MEDIUM,
    marginHorizontal: 20,
    marginVertical: 10,
  },
  imageWrapper: {
    borderRadius: 5,
    ...Platform.select({
      ios: {
        shadowColor: 'black',
        shadowOffset: { width: 3, height: 3 },
        shadowOpacity: 0.4,
        shadowRadius: 10,
      },
      android: {
        elevation: 5,
      },
    }),
  },
  image: {
    width: config.imageWidth,
    height: config.imageHeight,
    resizeMode: 'contain',
    borderRadius: 5,
    marginBottom: 30,
  },
});

export default ImageCard;
