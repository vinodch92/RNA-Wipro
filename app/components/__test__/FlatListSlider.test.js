import React from 'react';
import FlatListSlider from '../FlatListSlider';
import ImageCard from '../ImageCard';
import { data } from '../../../__fixture__/imagesSliderData';
import { render } from '@testing-library/react-native';
import { Dimensions } from 'react-native';

describe('FlatListSlider', () => {
  /* returns a component to render in flatlist */
  const renderItem = ({ item }) => {
    return (
      <ImageCard
        id={item.id}
        authorName={item.author}
        textNumLines={1}
        containerStyle={{
          marginHorizontal: 20,
          width: Math.round(Dimensions.get('screen').width) - 40,
        }}
      />
    );
  };

  const props = {
    data,
    renderItem: renderItem,
  };

  const dataLength = data.length;

  it('matches snapshot', () => {
    expect(render(<FlatListSlider />)).toMatchSnapshot();
    expect(render(<FlatListSlider {...props} dataLength={dataLength} />)).toMatchSnapshot();
  });
});
