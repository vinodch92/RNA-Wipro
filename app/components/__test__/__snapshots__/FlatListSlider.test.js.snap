// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlatListSlider matches snapshot 1`] = `
<RCTScrollView
  bounces={false}
  contentContainerStyle={
    Array [
      Object {
        "alignItems": "center",
      },
      undefined,
    ]
  }
  decelerationRate="fast"
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  horizontal={true}
  initialNumToRender={1}
  keyExtractor={[Function]}
  maxToRenderPerBatch={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  pagingEnabled={true}
  removeClippedSubviews={true}
  renderItem={[Function]}
  scrollEventThrottle={50}
  showsHorizontalScrollIndicator={false}
  snapToInterval={750}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={1}
>
  <View />
</RCTScrollView>
`;

exports[`FlatListSlider matches snapshot 2`] = `
<RCTScrollView
  bounces={false}
  contentContainerStyle={
    Array [
      Object {
        "alignItems": "center",
      },
      undefined,
    ]
  }
  data={
    Array [
      Object {
        "author": "Alejandro Escamilla",
        "author_url": "https://unsplash.com/photos/yC-Yzbqy7PY",
        "filename": "0.jpeg",
        "format": "jpeg",
        "height": 3744,
        "id": 0,
        "post_url": "https://unsplash.com/photos/yC-Yzbqy7PY",
        "width": 5616,
      },
      Object {
        "author": "Alejandro Escamilla",
        "author_url": "https://unsplash.com/photos/LNRyGwIJr5c",
        "filename": "1.jpeg",
        "format": "jpeg",
        "height": 3744,
        "id": 1,
        "post_url": "https://unsplash.com/photos/LNRyGwIJr5c",
        "width": 5616,
      },
      Object {
        "author": "Paul Jarvis",
        "author_url": "https://unsplash.com/photos/6J--NXulQCs",
        "filename": "10.jpeg",
        "format": "jpeg",
        "height": 1667,
        "id": 10,
        "post_url": "https://unsplash.com/photos/6J--NXulQCs",
        "width": 2500,
      },
      Object {
        "author": "Tina Rataj",
        "author_url": "https://unsplash.com/photos/pwaaqfoMibI",
        "filename": "100.jpeg",
        "format": "jpeg",
        "height": 1656,
        "id": 100,
        "post_url": "https://unsplash.com/photos/pwaaqfoMibI",
        "width": 2500,
      },
    ]
  }
  decelerationRate="fast"
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  horizontal={true}
  initialNumToRender={1}
  keyExtractor={[Function]}
  maxToRenderPerBatch={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  pagingEnabled={true}
  removeClippedSubviews={true}
  renderItem={[Function]}
  scrollEventThrottle={50}
  showsHorizontalScrollIndicator={false}
  snapToInterval={750}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={1}
>
  <View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
            },
            Object {
              "marginHorizontal": 20,
              "width": 710,
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "fontSize": 16,
                "fontWeight": "500",
                "marginHorizontal": 20,
                "marginVertical": 10,
              },
              undefined,
            ]
          }
          testID="author-name-text"
        >
          Alejandro Escamilla
        </Text>
        <View
          style={
            Array [
              Object {
                "borderRadius": 5,
                "shadowColor": "black",
                "shadowOffset": Object {
                  "height": 3,
                  "width": 3,
                },
                "shadowOpacity": 0.4,
                "shadowRadius": 10,
              },
              undefined,
            ]
          }
        >
          <Image
            source={
              Object {
                "uri": "https://picsum.photos/200/300?image=0",
              }
            }
            style={
              Object {
                "borderRadius": 5,
                "height": 300,
                "marginBottom": 30,
                "resizeMode": "contain",
                "width": 200,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "width": 2250,
        }
      }
    />
  </View>
</RCTScrollView>
`;
